### PR TITLE
[SPARK-13904][Scheduler]Add support for pluggable cluster manager 

### PR DESCRIPTION
**Pull request by hbhanawat**
## What changes were proposed in this pull request?

This commit adds support for pluggable cluster manager. And also allows a cluster manager to clean up tasks without taking the parent process down.

To plug a new external cluster manager, ExternalClusterManager trait should be implemented. It returns task scheduler and backend scheduler that will be used by SparkContext to schedule tasks. An external cluster manager is registered using the java.util.ServiceLoader mechanism (This mechanism is also being used to register data sources like parquet, json, jdbc etc.). This allows auto-loading implementations of ExternalClusterManager interface.

Currently, when a driver fails, executors exit using system.exit. This does not bode well for cluster managers that would like to reuse the parent process of an executor. Hence,
1. Moving system.exit to a function that can be overriden in subclasses of CoarseGrainedExecutorBackend.
2. Added functionality of killing all the running tasks in an executor.
## How was this patch tested?

ExternalClusterManagerSuite.scala was added to test this patch.

### CODE REVIEW

**Comment by sarutak** [comment]
ok to test.

**Comment by rxin** [comment]
Hmm I'm not sure if it makes sense to create a public API for this at this point. This is directly exposing something that is very internal to the current implementation of Spark (SchedulerBackend), and these are by no means stable.

**Comment by hbhanawat** [comment]
@rxin  Thanks for commenting. 

Spark was designed such that it is agnostic to the underlying cluster manager (as long as it can acquire executor processes, and these communicate with each other). Since Spark is now being used in newer and different use cases, there is a need for allowing other cluster managers to manage spark components. One such use case is - embedding spark components like executor and driver inside another process which may be a datastore. This allows co-location of data and processing. Another use case would be using Spark like an application server (you might have heard about spark-jobserver). Spark's current design allows handling such use cases if the cluster manager supports it. Hence, IMO, it is meaningful to allow plugging in new cluster managers. 

From code perspective, I think that even creation of TaskScheduler and SchedulerBackend for Yarn/Mesos/local mode should be done using a similar interface.

**Comment by rxin** [comment]
@hbhanawat I understand that. The problem is not whether you can find a single legitimate use case.

The introduction of every API always benefit something -- there is no argument about it. Otherwise nobody would be adding new APIs. The question is how many it benefits, and how the APIs can be maintained and evolved. You are effectively taking a bunch of private APIs that were never meant to be public and making them public. This approach is not maintainable.

**Comment by rxin** [comment]
@hbhanawat to be clear, I think we might be able to add this as a semi-private API and external resource managers can use, but with the understanding that this is tied to specific versions of Spark and might break from release over release. Exposing this as a stable public API based on my experience seeing how Spark has evolved in the last 5 years is not going to work.

**Comment by hbhanawat** [comment]
@rxin ok, I get it. I would make ExternalClusterManager as private[spark] and mark it as developer API. I hope that should suffice.

**Comment by hbhanawat** [comment]
@rxin I have completed the changes. Please review.

**Comment by hbhanawat** [comment]
@rxin Any update? Any changes needed from my side?

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 1]
```diff
@@ -2443,8 +2443,34 @@ object SparkContext extends Logging {
...
+
+  private def getClusterManager(url: String): Option[ExternalClusterManager] = {
+    val loader = Utils.getContextOrSparkClassLoader
+    val serviceLoader = ServiceLoader.load(classOf[ExternalClusterManager], loader)
+
+    serviceLoader.asScala.filter(_.canCreate(url)).toList match {
+      // exactly one registered manager
+      case head :: Nil => Some(head)
+      case Nil => None
+      case multipleMgrs => sys.error(s"Multiple Cluster Managers registered " +
```
Comment (tejasapatil) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: Can you include the list of matching cluster managers in the message ?

File: core/src/main/scala/org/apache/spark/executor/Executor.scala [hunk 2]
```diff
@@ -149,7 +149,14 @@ private[spark] class Executor(
       tr.kill(interruptThread)
     }
   }
-
+  def killAllTasks (interruptThread: Boolean) : Unit = {
```
Comment (tejasapatil) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: add a space between the methods

File: core/src/main/scala/org/apache/spark/executor/Executor.scala [hunk 3]
```diff
@@ -149,7 +149,14 @@ private[spark] class Executor(
...
+    // kill all the running tasks
+    for (taskRunner <- runningTasks.values().asScala) {
+      if (taskRunner != null) {
+        taskRunner.kill(interruptThread)
+      }
+    }
+  }
```
Comment (tejasapatil) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: add a space between the methods

File: core/src/main/scala/org/apache/spark/executor/Executor.scala (hunk 2 above)
Comment (tejasapatil) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: I could not see this method being called from anywhere. If you don't plan to use it, please remove this.

File: core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+package org.apache.spark.scheduler
+
+import org.apache.spark.SparkContext
+import org.apache.spark.annotation.DeveloperApi
+
+/**
+ * :: DeveloperApi ::
+ * A cluster manager interface to plugin external scheduler.
+ *
```
Comment (tejasapatil) [inline core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala]: remove extra space

File: core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    val conf = new SparkConf().setMaster("myclusterManager").
+        setAppName("testcm").set("spark.driver.allowMultipleContexts", "true")
+    sc = new SparkContext(conf)
+    // check if the scheduler components are created
+    assert(sc.schedulerBackend.isInstanceOf[FakeSchedulerBackend])
+    assert(sc.taskScheduler.isInstanceOf[FakeScheduler])
+  }
+}
+
+class CheckExternalClusterManager extends ExternalClusterManager {
```
Comment (tejasapatil) [inline core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala]: Rename to `TestExternalClusterManager` or `DummyExternalClusterManager`

File: core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+   * @return True if the cluster manager can create scheduler backend/
+   */
+  def canCreate(masterURL : String): Boolean
+
+  /**
+   * Create a task scheduler instance for the given SparkContext
+   * @param sc SparkContext
+   * @return TaskScheduler that will be responsible for task handling
+   */
+  def createTaskScheduler (sc: SparkContext): TaskScheduler
```
Comment (tejasapatil) [inline core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala]: can you also include the `masterURL` as a param ?

File: core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+  /**
+   * Create a scheduler backend for the given SparkContext and scheduler. This is
+   * called after task scheduler is created using [[ExternalClusterManager.createTaskScheduler()]].
+   * @param sc SparkContext
+   * @param scheduler TaskScheduler that will be used with the scheduler backend.
+   * @return SchedulerBackend that works with a TaskScheduler
+   */
+  def createSchedulerBackend (sc: SparkContext, scheduler: TaskScheduler): SchedulerBackend
```
Comment (tejasapatil) [inline core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala]: can you also include the `masterURL` as a param ?

File: core/src/test/resources/META-INF/services/org.apache.spark.scheduler.ExternalClusterManager [hunk 8]
```diff
@@ -0,0 +1 @@
+org.apache.spark.scheduler.CheckExternalClusterManager
```
Comment (tejasapatil) [inline core/src/test/resources/META-INF/services/org.apache.spark.scheduler.ExternalClusterManager]: why is this file needed ?

File: dev/.rat-excludes [hunk 9]
```diff
@@ -98,3 +98,4 @@ LZ4BlockInputStream.java
 spark-deps-.*
 .*csv
 .*tsv
+org.apache.spark.scheduler.ExternalClusterManager
```
Comment (tejasapatil) [inline dev/.rat-excludes]: why is this needed ?

File: core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  def createTaskScheduler(sc: SparkContext): TaskScheduler = new FakeScheduler
+
+  def createSchedulerBackend(sc: SparkContext, scheduler: TaskScheduler): SchedulerBackend =
+    new FakeSchedulerBackend()
+
+  def initialize(scheduler: TaskScheduler, backend: SchedulerBackend): Unit = {}
+
+}
+
+class FakeScheduler extends TaskScheduler {
```
Comment (tejasapatil) [inline core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala]: To keep this consistent with the external CM declared above, you could rename this to `TestTaskScheduler` or `DummyTaskScheduler`

**Comment by tejasapatil** [comment]
@rxin : I would really like to have this PR in trunk. As things stand, for anyone using their own scheduler, one has to maintain a patch over open source release to have that glue in Spark. Re API breaking over Spark releases : I agree that breaking APIs is bad. But it would be atleast better than the current model of dealing with this: doing a merge for *every* release.

**Comment by rxin** [comment]
Is this something Facebook needs too?

**Comment by tejasapatil** [comment]
@rxin : Yes !! At Facebook we are using an internal scheduler to run Spark executors. Maintaining an internal patch to have that "glue" and merging it against every Spark release can be avoided by this PR. Ideally we want to get to a place where there are no patches to maintain and everything is just pluggable. If we get there, testing RCs would be easier and we can help flagging issues early on (especially ones related to scalability).

**Comment by rxin** [comment]
OK - once you are done with your own review ping me. I will take a look at it again.

File: core/src/main/scala/org/apache/spark/SparkContext.scala (hunk 1 above)
Comment (hbhanawat) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: Done.

File: core/src/main/scala/org/apache/spark/executor/Executor.scala (hunk 2 above)
Comment (hbhanawat) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: 1. Done 
2. We are targeting a use case where in executors are launched inside another running process and when an executor goes down, it does not take the parent parent process down. In such cases, executors should kill the running tasks when they go down.  Given that, the runningtasks a private val, we need a method that can be called from the executor backend.

File: core/src/main/scala/org/apache/spark/executor/Executor.scala (hunk 3 above)
Comment (hbhanawat) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: Done

File: core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala (hunk 4 above)
Comment (hbhanawat) [inline core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala]: Done.

File: core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala (hunk 5 above)
Comment (hbhanawat) [inline core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala]: Done.

File: core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala (hunk 6 above)
Comment (hbhanawat) [inline core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala]: I have added the master url. In the hindsight, I think it makes sense.

File: core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala (hunk 7 above)
Comment (hbhanawat) [inline core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala]: Done.

File: core/src/test/resources/META-INF/services/org.apache.spark.scheduler.ExternalClusterManager (hunk 8 above)
Comment (hbhanawat) [inline core/src/test/resources/META-INF/services/org.apache.spark.scheduler.ExternalClusterManager]: To instantiate the dummy cluster manager in the test.

File: dev/.rat-excludes (hunk 9 above)
Comment (hbhanawat) [inline dev/.rat-excludes]: Rat throws an error for the files that do not have apache license. Hence excluding this file.

Comment (tejasapatil) [inline dev/.rat-excludes]: The version of `ExternalClusterManager` in the PR has a license header in it. Can you remove this and try ?

File: core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+import org.apache.spark.storage.BlockManagerId
+
+class ExternalClusterManagerSuite extends SparkFunSuite with LocalSparkContext
+{
+  test("launch of backend and scheduler") {
+    val conf = new SparkConf().setMaster("myclusterManager").
+        setAppName("testcm").set("spark.driver.allowMultipleContexts", "true")
+    sc = new SparkContext(conf)
+    // check if the scheduler components are created
+    assert(sc.schedulerBackend.isInstanceOf[FakeSchedulerBackend])
```
Comment (tejasapatil) [inline core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala]: Sorry I missed this in the last review comments yesterday. I thought that `FakeSchedulerBackend` was in this same file and you could rename it but now I see that its from some other place. 

While reading, it feels odd to have Fake* and then Dummy* test classes. I am not sure about the whats followed in Spark codebase. Couple options:
- rename Dummy* classes => Fake*. Move all the Fake* classes to a common test utils file for the module.
- Instead of re-using `FakeSchedulerBackend` from another place, create a `FakeSchedulerBackend` here.

File: core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    assert(sc.taskScheduler.isInstanceOf[DummyTaskScheduler])
+  }
+}
+
+class DummyExternalClusterManager extends ExternalClusterManager {
```
Comment (tejasapatil) [inline core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala]: `private`

File: core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+  def createSchedulerBackend(sc: SparkContext,
+                             masterURL: String,
+                             scheduler: TaskScheduler): SchedulerBackend =
+    new FakeSchedulerBackend()
+
+  def initialize(scheduler: TaskScheduler, backend: SchedulerBackend): Unit = {}
+
+}
+
+class DummyTaskScheduler extends TaskScheduler {
```
Comment (tejasapatil) [inline core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala]: `private`

File: dev/.rat-excludes (hunk 9 above)
Comment (hbhanawat) [inline dev/.rat-excludes]: I believe there is some confusion.  There is a file named org.apache.spark.scheduler.ExternalClusterManager in a test folder that is used by the ServiceLoader to launch the dummy external cluster manager. This file does not (and cannot) have a header and hence it needs to be excluded.

File: core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala (hunk 11 above)
Comment (hbhanawat) [inline core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala]: I too  missed it completely.  
I think it wasn't a great idea in the first place to use FakeSchedulerBackend of some other class from maintenance perspective. I am going ahead with your option 2.

File: core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala (hunk 12 above)
Comment (hbhanawat) [inline core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala]: Done.

File: core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala (hunk 13 above)
Comment (hbhanawat) [inline core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala]: Done

**Comment by hbhanawat** [comment]
@rxin  @tejasapatil  the previous build failure is not related to my checkin  and looks like some other issue as other build requests failed with the same exception. In case, it fails again, how do I restart it?

**Comment by hbhanawat** [comment]
@tejasapatil if you are done with the review, can we ask @rxin to have a look at this?

**Comment by tejasapatil** [comment]
sure. @rxin can you please review the PR ?

**Comment by tejasapatil** [comment]
Can someone review this PR ?

**Comment by hbhanawat** [comment]
@rxin can you please review this PR?

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala [hunk 14]
```diff
@@ -135,6 +135,8 @@ private[spark] class CoarseGrainedExecutorBackend(
       case None => logWarning(s"Drop $msg because has not yet connected to driver")
     }
   }
+
+  def exitExecutor(): Unit = System.exit(1)
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: can you add some document here saying why this is a public API - otherwise somebody might come in and remove this next time this code gets used.

File: core/src/main/scala/org/apache/spark/executor/Executor.scala [hunk 15]
```diff
@@ -150,6 +150,15 @@ private[spark] class Executor(
     }
   }
 
+  def killAllTasks (interruptThread: Boolean) : Unit = {
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: same here -- say why this is a public thing and how it is used.

also remove the space after killAllTasks

File: core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+
+import org.apache.spark.SparkContext
+import org.apache.spark.annotation.DeveloperApi
+
+/**
+  * :: DeveloperApi ::
+  * A cluster manager interface to plugin external scheduler.
+  */
+@DeveloperApi
+private[spark] trait ExternalClusterManager {
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala]: are you sure you want this to be private spark?

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 17]
```diff
@@ -2443,8 +2443,34 @@ object SparkContext extends Logging {
...
+          case Some(clusterMgr) => clusterMgr
+          case None => throw new SparkException("Could not parse Master URL: '" + master + "'")
+        }
+        try {
+          val scheduler = cm.createTaskScheduler(sc, masterUrl)
+          val backend = cm.createSchedulerBackend(sc, masterUrl, scheduler)
+          cm.initialize(scheduler, backend)
+          (backend, scheduler)
+        } catch {
+          case e: Exception => {
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: catch NonFatal here, i.e.
```
case NonFatal(e) =>
```

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 18]
```diff
@@ -2443,8 +2443,34 @@ object SparkContext extends Logging {
...
+            throw new SparkException("External scheduler cannot be instantiated", e)
+          }
+        }
+    }
+  }
+
+  private def getClusterManager(url: String): Option[ExternalClusterManager] = {
+    val loader = Utils.getContextOrSparkClassLoader
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: maybe
```scala
val loader = Utils.getContextOrSparkClassLoader
val serviceLoaders = ServiceLoader.load(classOf[ExternalClusterManager], loader).asScala.filter(_.canCreate(url))
if (serviceLoaders.size > 1) {
  throw new SparkException("Multiple Cluster Managers ($multipleMgrs) registered for the URL $url")
}
serviceLoaders.headOption
```

File: core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+
+  /**
+    * Check if this cluster manager instance can create scheduler components
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala]: use javadoc style indent in spark, i.e.

```
/**
 * Check if this cluster manager instance can create scheduler components
 * ...
 */
```

File: core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+
+  /**
+    * Create a scheduler backend for the given SparkContext and scheduler. This is
+    * called after task scheduler is created using [[ExternalClusterManager.createTaskScheduler()]].
+    * @param sc SparkContext
+    * @param masterURL the master URL
+    * @param scheduler TaskScheduler that will be used with the scheduler backend.
+    * @return SchedulerBackend that works with a TaskScheduler
+    */
+  def createSchedulerBackend(sc: SparkContext,
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala]: the way we indent is
```
def createSchedulerBackend(
    sc: SparkContext,
    masterURL: String,
    scheduler: TaskScheduler): SchedulerBackend
```

File: core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+  def defaultParallelism(): Int = 1
+}
+
+private class DummyTaskScheduler extends TaskScheduler {
+  override def rootPool: Pool = null
+  override def schedulingMode: SchedulingMode = SchedulingMode.NONE
+  override def start(): Unit = {}
+  override def stop(): Unit = {}
+  override def submitTasks(taskSet: TaskSet): Unit = {}
+  override def cancelTasks(stageId: Int, interruptThread: Boolean) {}
```
Comment (rxin) [inline core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala]: add a Unit here for return type to be consistent

File: core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+    assert(sc.schedulerBackend.isInstanceOf[DummySchedulerBackend])
+    assert(sc.taskScheduler.isInstanceOf[DummyTaskScheduler])
+  }
+}
+
+private class DummyExternalClusterManager extends ExternalClusterManager {
+
+  def canCreate(masterURL: String): Boolean = masterURL == "myclusterManager"
+
+  def createTaskScheduler(sc: SparkContext,
```
Comment (rxin) [inline core/src/test/scala/org/apache/spark/scheduler/ExternalClusterManagerSuite.scala]: can you fix the indent for the functions?

**Comment by rxin** [comment]
Sorry for the delay. This looks pretty good. Just have some comments about the style to be more consistent with rest of the Spark codebase.

**Comment by rxin** [comment]
there are some conflicts with master - can you rebase? Thanks.

**Comment by rxin** [comment]
The latest changes LGTM

**Comment by hbhanawat** [comment]
test this please

**Comment by hbhanawat** [comment]
@rxin how do I get this retested by Jenkins? There were few issues going on with the Jenkins when I checked in my last changes and now it is not retesting it?

**Comment by srowen** [comment]
Jenkins retest this please

**Comment by srowen** [comment]
Jenkins add to whitelist

File: core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+
+import org.apache.spark.SparkContext
+import org.apache.spark.annotation.DeveloperApi
+
+/**
+ * :: DeveloperApi ::
+ * A cluster manager interface to plugin external scheduler.
+ */
+@DeveloperApi
+trait ExternalClusterManager {
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/scheduler/ExternalClusterManager.scala]: i just realized most of the return types used in this class are private[spark], so your implementation of this interface would need to be in the spark package anyway. I'm going to add private[spark] to this when I merge.

**Comment by rxin** [comment]
Merging in master. Thanks.

**Comment by rxin** [comment]
One thing - can you guys try to see if you can implement one of the existing cluster manager with this, and then we can make sure this is a proper API? Otherwise it is really easy to get removed because it is currently unused by anything in Spark.

**Comment by hbhanawat** [comment]
@rxin I will open another JIRA and a PR to do this. Thanks for the review.

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala [hunk 24]
```diff
@@ -140,6 +140,13 @@ private[spark] class CoarseGrainedExecutorBackend(
       case None => logWarning(s"Drop $msg because has not yet connected to driver")
     }
   }
+
+  /**
+   * This function can be overloaded by other child classes to handle
+   * executor exits differently. For e.g. when an executor goes down,
+   * back-end may not want to take the parent process down.
+   */
+  protected def exitExecutor(): Unit = System.exit(1)
```
Comment (tedyu) [inline core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: Should a parameter of exit code (int) be added to this method ?

Comment (tedyu) [inline core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: Created https://github.com/apache/spark/pull/12457
